### PR TITLE
[DirectX] Add resource handling to the DXIL pretty printer

### PR DESCRIPTION
**Pull request by bogner**
Handle target extension type resources when printing resources to textual IR.

### CODE REVIEW

**Comment by bogner** [comment]
Depends on #104446

**Review by coopp — APPROVED** [review]
Looks good to me.  This is the first time I have seen a switch 'default:' come before the 'case:' entries.  (void format(llvm::raw_ostream &OS, StringRef Style) override)

File: llvm/test/CodeGen/DirectX/CreateHandle.ll [hunk 1]
```diff
@@ -1,4 +1,13 @@
 ; RUN: opt -S -passes=dxil-op-lower,dxil-translate-metadata %s | FileCheck %s
+; RUN: opt -S -passes=dxil-pretty-printer %s 2>&1 >/dev/null | FileCheck --check-prefix=CHECK-PRETTY %s
+
+; CHECK-PRETTY:       Type  Format         Dim      ID      HLSL Bind     Count
+; CHECK-PRETTY: ---------- ------- ----------- ------- -------------- ---------
+; CHECK-PRETTY:        UAV     f32         buf      U0      u5,space3         1
+; CHECK-PRETTY:        UAV     i32         buf      U1      u7,space2         1
+; CHECK-PRETTY:        SRV     u32         buf      T0      t3,space5        24
+; CHECK-PRETTY:        SRV  struct         r/o      T1      t2,space4         1
+; CHECK-PRETTY:        SRV    byte         r/o      T2      t8,space1         1
```
Comment (hekota) [inline llvm/test/CodeGen/DirectX/CreateHandle.ll:11]: Change one of the cases to use the default `space0`?

File: llvm/lib/Target/DirectX/DXILPrettyPrinter.cpp [hunk 2]
```diff
@@ -10,23 +10,235 @@
...
+           "Old and new UAV representations can't coexist");
+
+    StringRef Name(RI.getName());
+    StringRef Type(getRCName(RC));
+    StringRef Format(getFormatName(RI));
+    FormatResourceDimension Dim(RI);
+    FormatBindingID ID(RI);
+    FormatBindingLocation Bind(RI);
+    FormatBindingSize Count(RI);
+    OS << formatv("; {0,-30} {1,10} {2,7} {3,11} {4,7} {5,14} {6,9}\n",
```
Comment (hekota) [inline llvm/lib/Target/DirectX/DXILPrettyPrinter.cpp:238]: Neat use of the ` llvm::FormatAdapter`. Bit of a overkill for the simple cases but it looks elegant. ;).

**Review by hekota — APPROVED** [review]
LGTM!

**Review by damyanp — APPROVED** [review]

File: llvm/lib/Target/DirectX/DXILPrettyPrinter.cpp [hunk 3]
```diff
@@ -10,23 +10,235 @@
...
+#include "llvm/InitializePasses.h"
 #include "llvm/Pass.h"
+#include "llvm/Support/FormatAdapters.h"
 #include "llvm/Support/FormatVariadic.h"
 #include "llvm/Support/raw_ostream.h"
 
 using namespace llvm;
 
-static void prettyPrintResources(raw_ostream &OS,
+static constexpr StringRef getRCName(dxil::ResourceClass RC) {
```
Comment (farzonl) [inline llvm/lib/Target/DirectX/DXILPrettyPrinter.cpp:23]: Feel free to ignore, I was thinking of a different  way to do this that would have a tighter coupling of  Names and prefixes:

```cpp
struct ResourceClassInfo {
  const StringRef name;
  const StringRef prefix;
};

llvm::DenseMap<dxil::ResourceClass, ResourceClassInfo> createResourceClassMap() {
  return {
    {dxil::ResourceClass::SRV, {"SRV", "t"}},
    {dxil::ResourceClass::UAV, {"UAV", "u"}},
    {dxil::ResourceClass::CBuffer, {"cbuffer", "cb"}},
    {dxil::ResourceClass::Sampler, {"sampler", "s"}}
  };
}

static const llvm::DenseMap<dxil::ResourceClass, ResourceClassInfo> ResourceClassMap = createResourceClassMap();

StringRef getRCName(dxil::ResourceClass RC) {
  return ResourceClassMap.lookup(RC).name;
}

StringRef getRCPrefix(dxil::ResourceClass RC) {
  return ResourceClassMap.lookup(RC).prefix;
}
```

**Review by farzonl — APPROVED** [review]
LGTM

**Pull request merged**